nat/nat_status_v2.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2021 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package nat
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/core/node"
+
+	"github.com/mysteriumnetwork/node/core"
+
+	"github.com/mysteriumnetwork/node/config"
+	"github.com/mysteriumnetwork/node/identity"
+)
+
+// NATStatusV2 enum
+// TODO: V2 suffix should be removed once previous tracker is removed
+type NATStatusV2 string
+
+const (
+	// Online enum
+	Online NATStatusV2 = "online"
+	// Offline enum
+	Offline NATStatusV2 = "offline"
+	// Failed enum
+	Failed NATStatusV2 = "failed"
+	// Pending enum
+	Pending NATStatusV2 = "pending"
+)
+
+// AppTopicNATStatusUpdate nat status update topic for event bus
+const AppTopicNATStatusUpdate = "AppTopicNATStatusUpdate"
+
+type currentIdentity interface {
+	GetUnlockedIdentity() (identity.Identity, bool)
+}
+
+type publisher interface {
+	Publish(topic string, data interface{})
+}
+
+// ProviderSessions should return provider session monitoring state
+type ProviderSessions func(providerID string) []Session
+
+// Session represent session monitoring state
+type Session struct {
+	ProviderID       string
+	ServiceType      string
+	MonitoringFailed bool
+}
+
+// V2NatStatusEvent nat status event
+type V2NatStatusEvent struct {
+	Status NATStatusV2
+}
+
+// StatusTrackerV2 tracks NAT status for service
+type StatusTrackerV2 struct {
+	lock   sync.Mutex
+	status NATStatusV2
+
+	publisher publisher
+
+	providerSessions ProviderSessions
+	currentIdentity  currentIdentity
+
+	pollInterval time.Duration
+}
+
+// NewStatusTrackerV2 constructor
+func NewStatusTrackerV2(
+	providerSessions ProviderSessions,
+	currentIdentity currentIdentity,
+	publisher publisher,
+	options node.OptionsNATStatusTrackerV2,
+) *StatusTrackerV2 {
+	validatedInterval := options.PollInterval
+	if validatedInterval < time.Minute {
+		validatedInterval = time.Minute
+	}
+	keeper := &StatusTrackerV2{
+		providerSessions: providerSessions,
+		currentIdentity:  currentIdentity,
+		publisher:        publisher,
+		pollInterval:     validatedInterval,
+	}
+	// consumers don't need service NAT status
+	if !config.GetBool(config.FlagConsumer) {
+		keeper.startPolling()
+	}
+	return keeper
+}
+
+func (k *StatusTrackerV2) startPolling() {
+	k.updateAndAnnounce()
+	go func() {
+		for {
+			select {
```

Comment:
No channels to select from so the `select` is not really needed here

Reply:
or even just use time.Sleep